### PR TITLE
CP-44372: Integrate NRPE UI with backend interface

### PR DESCRIPTION
**Pull request by BengangY**
Added below use case on XenCenter

Enable/Disable NRPE for host
Configure allow_hosts /Enable Debug/Enable SSL_logging
Set threshold for each check
View existing settings for each host
Validation for the input
Same function for the pool

### CODE REVIEW

File: XenAdmin/Dialogs/PropertiesDialog.cs [hunk 1]
```diff
@@ -317,6 +318,11 @@ private void Build()
                         dialog.ShowDialog(Program.MainWindow);
                     }
                 }
+                if (isHost || isPool)
```
Comment (kc284) [inline XenAdmin/Dialogs/PropertiesDialog.cs:321]: The page should be available only if the user connects to servers that have the NRPE plugin installed. XC is allowed to connect to older servers which do not have the plugin, so we need to somehow check and hide the page.

File: XenAdmin/SettingsPanels/NRPEEditPage.cs [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+                InvalidParamToolTip.Hide(ctrl);
+            }
+        }
+
+        public void SetXenObjects(IXenObject orig, IXenObject clone)
+        {
+            Clone = clone;
+            if (IsHost)
+            {
+                InitNRPEGeneralConfiguration();
```
Comment (kc284) [inline XenAdmin/SettingsPanels/NRPEEditPage.cs:206]: These two methods make server calls on the UI thread, which means that if the network is slow the UI will freeze. Please create a new AsyncAction in the XenModel/Actions/NRPE folder (see my commits in https://github.com/kc284/xenadmin/commits/private/bengangy/CP-44371) and move this code in there. This page should be running the async action, hook the Completed event and update the UI on success/failure accordingly, While the action is running, the UI should have some indication that something is happening on the background, so you may want to add a spinner panel saying something like "Retrieving NRPE configurarion..."

File: XenAdmin/SettingsPanels/NRPEEditPage.cs [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+                }
+                return valid;
+            }
+        }
+
+        public bool HasChanged
+        {
+            get
+            {
+                return true;
```
Comment (kc284) [inline XenAdmin/SettingsPanels/NRPEEditPage.cs:173]: It is not guaranteed that the user will change something. They may click OK without any changes in this dialog or they becasue they have changed something in a different panel. In this case we should not run the action to save changes unnecessarily

File: XenAdmin/SettingsPanels/NRPEEditPage.cs [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+            if (string.IsNullOrWhiteSpace(AllowHostsTextBox.Text))
+            {
+                AllowHostsTextBox.Text = NRPEHostConfiguration.ALLOW_HOSTS_PLACE_HOLDER;
+                AllowHostsTextBox.ForeColor = Color.FromKnownColor(KnownColor.ControlDark);
+            }
+        }
+
+        private void CheckDataGridView_BeginEdit(object sender, DataGridViewCellCancelEventArgs e)
+        {
+            DataGridViewCell currentCell = CheckDataGridView.CurrentRow.Cells[e.ColumnIndex];
```
Comment (kc284) [inline XenAdmin/SettingsPanels/NRPEEditPage.cs:377]: Here and in the following event handler I'm not convinced that checking the colour is the right way to decide to use thewatermark or not. It would be bette if we had a way to check whether the user has modified the value or not, in which case the cell should have the normal colour to make it obvious.

File: XenModel/Actions/Host/NRPEUpdateAction.cs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    {
+        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        private readonly NRPEHostConfiguration NRPEOriginalConfig;  // NRPE configuration fetched from host
+        private readonly NRPEHostConfiguration NRPEHostConfiguration;  // NRPE configuration after user modified
+
+        private readonly IXenObject Clone;
+
+        public NRPEUpdateAction(IXenObject host, NRPEHostConfiguration nrpeHostConfiguration, NRPEHostConfiguration nrpeOriginalConfig, bool suppressHistory)
+            : base(host.Connection, Messages.ACTION_CHANGE_POWER_ON, Messages.NRPE_ACTION_CHANGING, suppressHistory)
```
Comment (kc284) [inline XenModel/Actions/Host/NRPEUpdateAction.cs:48]: Copy paste error Messages.ACTION_CHANGE_POWER_ON,

**Review by kc284 — CHANGES_REQUESTED** [review]
As above.

**Comment by kc284** [comment]
A general remark: normally for new controls we would add localized resx placeholders, but at the moment we have stopped shipping localized resources on master, so the palceholders are not needed. This is the first PR with missing placeholders, so please ensure it is merged after #3238 which disables the unit tests.

File: XenAdmin/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -35,10 +35,10 @@
 // General Information about an assembly is controlled through the following 
 // set of attributes. Change these attribute values to modify the information
 // associated with an assembly.
-[assembly: AssemblyTitle("[XenCenter]")]
-[assembly: AssemblyDescription("[XenCenter]")]
+[assembly: AssemblyTitle("XenCenter")]
+[assembly: AssemblyDescription("XenCenter")]
 [assembly: AssemblyConfiguration("")]
-[assembly: AssemblyProduct("[XenCenter]")]
+[assembly: AssemblyProduct("XenCenter")]
```
Comment (danilo-delbusso) [inline XenAdmin/Properties/AssemblyInfo.cs:42]: Looks like these are leftover from building XenCenter with branding? 👀 We should revert them. Same for the value in `XenAdmin.csproj`

File: XenAdmin/SettingsPanels/NRPEEditPage.cs [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+                NRPEHostConfiguration.Check CurrentCheck = kvp.Value;
+                _nrpeOriginalConfig.GetNRPECheck(kvp.Key, out NRPEHostConfiguration.Check OriginalCheck);
+                if (CurrentCheck != null && OriginalCheck != null
+                    && (!CurrentCheck.WarningThreshold.Equals(OriginalCheck.WarningThreshold)
+                    || !CurrentCheck.CriticalThreshold.Equals(OriginalCheck.CriticalThreshold)))
+                {
+                    return true;
+                }
+            }
+            return false;
```
Comment (danilo-delbusso) [inline XenAdmin/SettingsPanels/NRPEEditPage.cs:434]: I'd put this logic in the `Equals` override for `NRPEHostConfiguration`, so they can be reused. then you can just use a nice and simple `Equals` here.

**Review by danilo-delbusso — CHANGES_REQUESTED** [review]
Haven't done a complete review but noticed a couple of things while skim-reading

**Review by kc284 — CHANGES_REQUESTED** [review]

File: XenAdmin/Dialogs/PropertiesDialog.cs [hunk 8]
```diff
@@ -317,6 +318,12 @@ private void Build()
                         dialog.ShowDialog(Program.MainWindow);
                     }
                 }
+                if (isPoolOrStandalone && Helpers.XapiEqualOrGreater_23_27_0(connection)
```
Comment (kc284) [inline XenAdmin/Dialogs/PropertiesDialog.cs:321]: Is `isPoolOrStandalone` correct here? This means that if someone wants to set different settings for only one host in a multi-host pool, they cannot do it.

**Comment by kc284** [comment]
- If I enter an invalid value in one of the threshold cells (e.g. 120), and click the OK button, I get an error (little red icon in the cell). If I then change the value to something valid (e.g. 70) and click on a different cell, the little red icon should disappear, but it's still there.
- If I enter an invalid value in a cell in the last row, click to a different tab, then click the OK button, it takes me back to the NRPE page, which is correct, however, the row with the invalid value is out of view, one has to scroll down to see what's wrong. Ideally, the row containing the invalid cell should be selected or at least we should scroll into view. If there are more than one row with invalid cells, it should be at least the first invalid row in view.

File: XenModel/Actions/NRPE/NRPEUpdateAction.cs [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+            // NRPE Check Threshold
+            foreach (KeyValuePair<string, NRPEHostConfiguration.Check> kvp in _nrpeHostConfiguration.CheckDict)
+            {
+                NRPEHostConfiguration.Check currentCheck = kvp.Value;
+                _nrpeOriginalConfig.GetNRPECheck(kvp.Key, out NRPEHostConfiguration.Check OriginalCheck);
+                if (currentCheck != null && OriginalCheck != null
+                    && (!currentCheck.WarningThreshold.Equals(OriginalCheck.WarningThreshold)
+                    || !currentCheck.CriticalThreshold.Equals(OriginalCheck.CriticalThreshold)))
+                {
+                    SetNRPEThreshold(o, currentCheck.Name, currentCheck.WarningThreshold, currentCheck.CriticalThreshold);
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/NRPEUpdateAction.cs:96]: This Host.call_plugin supports to set only one threshold at a time.
Host.call_plugin takes a parameter thresholdArgDict because checkName, warningThreshold, and criticalThreshold are 3 keys for this call.

File: XenModel/Actions/NRPE/NRPERetrieveAction.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            catch (Exception e)
+            {
+                log.ErrorFormat("Execute NRPE plugin failed, failed reason: {0}", e.Message);
+                _nrpeCurrentConfig.Status = NRPEHostConfiguration.RetrieveNRPEStatus.Failed;
+            }
+        }
+
+        private void InitNRPEGeneralConfiguration(IXenObject o)
+        {
+            string status = Host.call_plugin(o.Connection.Session, o.opaque_ref, NRPEHostConfiguration.XAPI_NRPE_PLUGIN_NAME,
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/NRPERetrieveAction.cs:75]: It's not .ini, json, or yaml file. It's a plain text structure, like this:

```
check_cpu warning threshold - 50 critical threshold - 80
check_host_memory warning threshold - 50 critical threshold - 80
check_vgpu_memory warning threshold - 50 critical threshold - 80
...
```
There is only NRPE plugin using this structure now. For it's hard to parse, new plugins in the future (e.g. SNMP) will use json structure instead.

File: XenAdmin/Dialogs/PropertiesDialog.cs (hunk 8 above)
Comment (BengangY) [reply XenAdmin/Dialogs/PropertiesDialog.cs:321]: It's correct. Have confirmed with Enzo that NRPE page will be shown for a pool and standalone host, not shown for a host in a multi-host pool.

File: XenModel/Actions/NRPE/CheckGroup.cs [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+        private string criticalThresholdDefault;
+        private bool changed;
+
+        protected DataGridViewRow checkThresholdRow;
+        protected DataGridViewTextBoxCell nameCell;
+        protected DataGridViewTextBoxCell warningThresholdCell;
+        protected DataGridViewTextBoxCell criticalThresholdCell;
+
+        public string Name { get => name; set => name = value; }
+        public string WarningThresholdDefault { get => warningThresholdDefault; set => warningThresholdDefault = value; }
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/CheckGroup.cs:55]: After refactoring, they are not used. Have removed them.

File: XenModel/Actions/NRPE/CheckGroup.cs [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+        {
+            InitCheckGroup(name, labelName, warningThresholdDefaultValue, criticalThresholdDefaultValue);
+        }
+
+        public CheckGroup(string name, string labelName)
+        {
+            InitCheckGroup(name, labelName, DEFAULT_CHECK_WARNING_THRESHOLD, DEFAULT_CHECK_CRITICAL_THRESHOLD);
+        }
+
+        private void InitCheckGroup(string name, string labelName, string warningThresholdDefaultValue, string criticalThresholdDefaultValue)
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/CheckGroup.cs:73]: Updated.

File: XenModel/Actions/NRPE/CheckGroup.cs [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+        public DataGridViewTextBoxCell NameCell { get => nameCell; set => nameCell = value; }
+        public DataGridViewTextBoxCell WarningThresholdCell { get => warningThresholdCell; set => warningThresholdCell = value; }
+        public DataGridViewTextBoxCell CriticalThresholdCell { get => criticalThresholdCell; set => criticalThresholdCell = value; }
+
+        public CheckGroup(string name, string labelName, string warningThresholdDefaultValue, string criticalThresholdDefaultValue)
+        {
+            InitCheckGroup(name, labelName, warningThresholdDefaultValue, criticalThresholdDefaultValue);
+        }
+
+        public CheckGroup(string name, string labelName)
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/CheckGroup.cs:68]: Updated.

File: XenModel/XenModel.csproj [hunk 14]
```diff
@@ -57,6 +57,7 @@
     <Reference Include="System.ServiceProcess" />
     <Reference Include="System.Web" />
     <Reference Include="System.Web.Extensions" />
+    <Reference Include="System.Windows.Forms" />
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/XenModel.csproj:60]: Updated.

File: XenModel/Actions/NRPE/NRPEUpdateAction.cs [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            Dictionary<string, string> thresholdArgDict = new Dictionary<string, string>
+                        {
+                            { checkName, null },
+                            { "w", warningThreshold },
+                            { "c", criticalThreshold }
+                        };
+            string result = Host.call_plugin(host.Connection.Session, host.opaque_ref, NRPEHostConfiguration.XAPI_NRPE_PLUGIN_NAME,
+                NRPEHostConfiguration.XAPI_NRPE_SET_THRESHOLD, thresholdArgDict);
+            log.InfoFormat("Execute nrpe {0}, check={1}, w={2}, c={3}, return: {4}",
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/NRPEUpdateAction.cs:149]: Updated.

File: XenModel/Actions/NRPE/NRPEUpdateAction.cs [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+                || !_nrpeHostConfiguration.SslLogging.Equals(_nrpeOriginalConfig.SslLogging))
+            {
+                SetNRPEGeneralConfiguration(o, _nrpeHostConfiguration.AllowHosts, _nrpeHostConfiguration.Debug, _nrpeHostConfiguration.SslLogging);
+            }
+
+            // NRPE Check Threshold
+            foreach (KeyValuePair<string, NRPEHostConfiguration.Check> kvp in _nrpeHostConfiguration.CheckDict)
+            {
+                NRPEHostConfiguration.Check currentCheck = kvp.Value;
+                _nrpeOriginalConfig.GetNRPECheck(kvp.Key, out NRPEHostConfiguration.Check OriginalCheck);
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/NRPEUpdateAction.cs:91]: Updated.

File: XenModel/Actions/NRPE/NRPEUpdateAction.cs [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+                }
+            }
+
+            RestartNRPE(o);
+        }
+
+        private void SetNRPEConfigureForPool()
+        {
+            List<Host> hostList = ((Pool) _clone).Connection.Cache.Hosts.ToList();
+            hostList.ForEach(host =>
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Actions/NRPE/NRPEUpdateAction.cs:106]: Updated.

File: XenModel/Messages.resx [hunk 18]
```diff
@@ -10092,6 +10092,99 @@ When you configure an NFS storage repository, you simply provide the host name o
...
+    <value>Host vGPU Usage (%)</value>
+  </data>
+  <data name="NRPE_EDIT_PAGE_TEXT" xml:space="preserve">
+    <value>NRPE Configuration</value>
+  </data>
+  <data name="NRPE_INACTIVE" xml:space="preserve">
+    <value>NRPE service is inactive</value>
+  </data>
+  <data name="NRPE_RETRIEVE_FAILED" xml:space="preserve">
+    <value>Retrieve NRPE configuration failed, please check XenCenter logs.</value>
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Messages.resx:10165]: Updated.

File: XenModel/Messages.resx [hunk 19]
```diff
@@ -10092,6 +10092,99 @@ When you configure an NFS storage repository, you simply provide the host name o
...
+  <data name="NRPE_ALLOW_HOSTS_ERROR_TITLE" xml:space="preserve">
+    <value>Monitoring servers format is not correct</value>
+  </data>
+  <data name="NRPE_ALLOW_HOSTS_FORMAT_ERROR" xml:space="preserve">
+    <value>Monitoring servers should be comma separated IP address or domain list, e.g. 192.168.1.1, test.domain.com</value>
+  </data>
+  <data name="NRPE_ALLOW_HOSTS_PLACE_HOLDER" xml:space="preserve">
+    <value>Comma separated IP address or domain list, e.g. 192.168.1.1, test.domain.com</value>
+  </data>
+  <data name="NRPE_ALLOW_HOSTS_SAME_ADDRESS" xml:space="preserve">
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenModel/Messages.resx:10119]: Updated.

File: XenAdmin/XenAdmin.csproj [hunk 20]
```diff
@@ -9,7 +9,7 @@
     <OutputType>WinExe</OutputType>
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>XenAdmin</RootNamespace>
-    <AssemblyName>[XenCenter]</AssemblyName>
+    <AssemblyName>XenCenter</AssemblyName>
```
[the remark this answers is not in the document]
Comment (BengangY) [reply XenAdmin/XenAdmin.csproj:12]: Updated.

**Comment by BengangY** [comment]
> * If I enter an invalid value in one of the threshold cells (e.g. 120), and click the OK button, I get an error (little red icon in the cell). If I then change the value to something valid (e.g. 70) and click on a different cell, the little red icon should disappear, but it's still there.
Updated. Now it will check when the datagridview lost focus each time.
> * If I enter an invalid value in a cell in the last row, click to a different tab, then click the OK button, it takes me back to the NRPE page, which is correct, however, the row with the invalid value is out of view, one has to scroll down to see what's wrong. Ideally, the row containing the invalid cell should be selected or at least we should scroll into view. If there are more than one row with invalid cells, it should be at least the first invalid row in view.
Updated. Added a reselection process when error occurs. Code likes:
```
CheckDataGridView.CurrentCell = CheckDataGridView.Rows[0].Cells[0];
CheckDataGridView.CurrentCell = focusCellWhenErrorOccurs;
```
It will select the first row and then select the error row. The reason is that if it only selects the error row, if the user already selected the error row and scrolled down the datagrid and click to a different tab, then click the OK button, it will not focus on the error row. So it needs a switch process.

File: XenModel/Actions/NRPE/NRPEUpdateAction.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (kc284) [reply XenModel/Actions/NRPE/NRPEUpdateAction.cs:98]: Ok. Maybe we can consider it as a future enhancement to the plugin.

File: XenAdmin/SettingsPanels/NRPEEditPage.cs [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+        {
+            if (_invalidParamToolTip.Tag is Control ctrl)
+            {
+                _invalidParamToolTip.Hide(ctrl);
+            }
+        }
+
+        public AsyncAction SaveSettings()
+        {
+            return new NRPEUpdateAction(_clone, _nrpeCurrentConfig, _nrpeOriginalConfig, true);
```
Comment (kc284) [inline XenAdmin/SettingsPanels/NRPEEditPage.cs:182]: I didn't spot this earlier: why are we suppressing the history? This means that there is not record of the action on the events TabPage. Same for line 195 and NRPERetrieveAction.

**Comment by kc284** [comment]
One more question: I noticed that when we disable NRPE and then we renable it, the previous settings are reset and the user has to set up thresholds again. Is this by design? (Asking because sometimes features do remember the settings when disabled).

File: XenAdmin/SettingsPanels/NRPEEditPage.cs (hunk 21 above)
Comment (kc284) [reply XenAdmin/SettingsPanels/NRPEEditPage.cs:182]: Hmm, actually we do suppress this for more properties that we save. It probably needs revisiting at a higher level, so it is probably a no-action for this PR.

**Comment by kc284** [comment]
I added a Thread.Sleep() for a few seconds within the Host.call_plugin() method to simulate what happens if, for example, we have a slow network and server requests take while to return. In this case, if I launch the dialog and click the NRPE page, I see an error pop-up:



Which means something is not working correctly if we click on the page while the NRPERetrieveAction is still in progress.

**Comment by BengangY** [comment]
> One more question: I noticed that when we disable NRPE and then we renable it, the previous settings are reset and the user has to set up thresholds again. Is this by design? (Asking because sometimes features do remember the settings when disabled).

What were your operation steps? There is a scenario, I'm not sure if this is your situation:
1 The current status of NRPE is enabled.
2 Modify some settings, e.g. thresholds, and disable NRPE.
3 Click OK.
Then XC will only send 'disable NRPE' to XS. It doesn't need to send the other settings because the NRPE has already been disabled. This s by design.

**Comment by BengangY** [comment]
> I added a Thread.Sleep() for a few seconds within the Host.call_plugin() method to simulate what happens if, for example, we have a slow network and server requests take while to return. In this case, if I launch the dialog and click the NRPE page, I see an error pop-up:
> 
> 
> Which means something is not working correctly if we click on the page while the NRPERetrieveAction is still in progress.

This is an issue caused by my removing the default values of the threshold in datagridview. I will fix it today.

**Review by kc284 — APPROVED** [review]

**Review by danilo-delbusso — APPROVED** [review]
